File: 33-search-in-rotated-sorted-array/33-search-in-rotated-sorted-array.java
class Solution {
    public int search(int[] nums, int target) {
        int pivot = findPivot(nums);
        if(target < nums[0])
            return find(nums,target,pivot + 1, nums.length - 1);
        else 
            return find(nums,target,0,pivot);
        
    }
    
    private int find(int[] nums, int target, int st, int end){
        while(st <= end){
            int mid = st + (end - st) / 2;
            if(nums[mid] > target) end = mid - 1;
            else if(nums[mid] < target) st = mid + 1;
            else return mid;
        }
        return -1;
    
    }
    
    private int findPivot(int[] nums){
        int st = 0; int end = nums.length - 1;
        
        while(st <= end){
            int mid = st + (end - st)/2;
            if(mid + 1 < nums.length && nums[mid+1] < nums[mid]) return mid;
            else if(nums[mid] < nums[0]) end = mid - 1;
            else st = mid + 1;
        }
        
        return (st == nums.length) ? st - 1 : st;
    }
}

File: 33-search-in-rotated-sorted-array/33-search-in-rotated-sorted-array.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        
        def find(target, st, end):
            idx = bisect_left(nums,target,st,end)
            if idx != end and target == nums[idx]:
                return idx 
            return -1 
        
            
        st, end = 0, len(nums) - 1 
        while st <= end:
            mid = (st + end ) // 2 
            if nums[mid] >= nums[0]: 
                st = mid + 1 
            else: 
                end = mid - 1 
        if st == len(nums):
            st -= 1 
        idx = find(target,st,len(nums))
        if idx != -1:
            return idx 
        return find(target,0, st)
